### PR TITLE
 [Snippets] Added dynamism support to MHA Tokenization pass

### PR DESCRIPTION
**Pull request by a-sidorova**
### Details:
 - *Updated `MHATokenization` pass to support dynamic nodes tokenization*
 - *Added MHATokenization unit tests*
 - *Disabled dynamic MHA tokenization in CPU Plugin*

### Tickets:
 - *123329*

### CODE REVIEW

File: src/common/snippets/src/pass/mha_tokenization.cpp [hunk 1]
```diff
@@ -224,20 +234,14 @@ ov::snippets::pass::TokenizeMHASnippets::TokenizeMHASnippets(const SnippetsToken
...
         //                                     But operations after Transpose1 and before MatMul0  will be fused into one loop as well (look at first point)
-        // Note: If the pass is updated, need to check the new possible branches for potential non-inplace Buffers!
-        // Default value is 2 because
-        //  - Firstly, Softmax always needs Buffers
-        //  - Secondly, Softmax needs 2 Buffers but they can be inplace - One virtual port is enough for Softmax => buffer_count = 1
-        //  - Thirdly, MatMul requires unique Buffers on inputs and outputs because blocking implementation increments input/output pointers during computations
-        //    However, all of the Buffers are usually reused by the next MatMul and Softmax.
-        //    So on sufficiently large subgraphs we use only one additional unique buffer => buffer_count increments by 1
-        size_t buffer_count = 2;
+        size_t uniqie_buffer_reg_group_count = 1;  // After MatMul0 there is always one Buffer
```
Comment (IvanNovoselov) [inline src/common/snippets/src/pass/mha_tokenization.cpp:244]: Why we previously needed 2 buffers, but now only one?
Is it because of the blocking support in dynamism?

File: src/common/snippets/src/pass/mha_tokenization.cpp [hunk 2]
```diff
@@ -260,24 +264,20 @@ ov::snippets::pass::TokenizeMHASnippets::TokenizeMHASnippets(const SnippetsToken
...
 
-        const auto matmul0_prc = op::Brgemm::get_output_type(matmul0->get_input_element_type(0), matmul0->get_input_element_type(1));
-        // Between MatMul0 and Softmax will be the one Loop because of LoopFusing optimization.
-        // The Loop will have one Buffer with the same shape both on input and output.
-        // Need to check for precision to get if we need one more register for Buffer
-        if (matmul0_prc.size() != ov::element::f32.size()) {
-            if (buffer_count < 2)
-                buffer_count++;
-        }
-
```
Comment (IvanNovoselov) [inline src/common/snippets/src/pass/mha_tokenization.cpp:271]: Did we remove this block because we are going to support only fp32 in dynamism for now?

File: src/common/snippets/src/pass/mha_tokenization.cpp [hunk 3]
```diff
@@ -294,6 +294,11 @@ ov::snippets::pass::TokenizeMHASnippets::TokenizeMHASnippets(const SnippetsToken
 
         if (axis != rank.get_length() - 1 || interm_op->get_output_target_inputs(0).size() != 1)
             return false;
+
+        // Softmax need one buffer at least
+        if (interm_op->get_output_partial_shape(0).rbegin()->is_dynamic())
```
Comment (IvanNovoselov) [inline src/common/snippets/src/pass/mha_tokenization.cpp:299]: What about static cases? Doesn't static Softmax require buffer as well?

File: src/plugins/intel_cpu/src/transformations/transformation_pipeline.cpp [hunk 4]
```diff
@@ -867,7 +867,7 @@ void Transformations::MainSnippets(void) {
 #if defined(OPENVINO_ARCH_X86_64)
     auto is_supported_matmul = [this](const std::shared_ptr<const ov::Node>& n) {
         const auto matmul = ov::as_type_ptr<const ov::op::v0::MatMul>(n);
-        if (!matmul)
+        if (!matmul || matmul->is_dynamic())
```
Comment (IvanNovoselov) [inline src/plugins/intel_cpu/src/transformations/transformation_pipeline.cpp:870]: Do I understand correctly that the tokenization tets we discussed offline can now be reproduced on this PR if we revert this change?

**Review by IvanNovoselov — COMMENTED** [review]
General comment: there are some changes that affect static tokenization in this PR. Have they been validated?

File: src/common/snippets/src/pass/mha_tokenization.cpp [hunk 5]
```diff
@@ -302,16 +307,15 @@ ov::snippets::pass::TokenizeMHASnippets::TokenizeMHASnippets(const SnippetsToken
...
-             (reshape0 && reshape1 && (reshape0->get_input_shape(0) != reshape1->get_output_shape(0))))
+             (reshape0 && reshape1 && (reshape0->get_input_partial_shape(0) != reshape1->get_output_partial_shape(0))))
             return false;
 
         // Add supported operations which are between Softmax and MatMul1 to ordered_ops
         if (!update_intermediate_supported_ops(interm_op, ordered_ops, hidden_virtual_ports_count, potential_body_params_count))
             return false;
 
         const auto matmul1 = ov::as_type_ptr<ov::opset1::MatMul>(interm_op);
-        if (!matmul1 || matmul1->get_output_target_inputs(0).size() != 1 ||
```
Comment (v-Golubev) [inline src/common/snippets/src/pass/mha_tokenization.cpp:313]: Why do we remove consumers count check? Because of ShapeOf subgraphs or smth else?

File: src/common/snippets/src/pass/mha_tokenization.cpp [hunk 6]
```diff
@@ -358,6 +363,7 @@ ov::snippets::pass::TokenizeMHASnippets::TokenizeMHASnippets(const SnippetsToken
         // There is transformation ExplicitTransposeMatMulInputs that set supported order and transposed_b(false).
         // We can allow to call this pass only if ops have scalar shapes to avoid shape mismatching
         const auto is_transposed_b_0 = matmul0->get_transpose_b();
+        bool has_matmul0_has_ops_on_input = false;
```
Comment (v-Golubev) [inline src/common/snippets/src/pass/mha_tokenization.cpp:366]: Minor typo
```suggestion
        bool matmul0_has_ops_on_input = false;
```

File: src/common/snippets/src/pass/mha_tokenization.cpp (hunk 1 above)
Comment (a-sidorova) [reply src/common/snippets/src/pass/mha_tokenization.cpp:244]: Because this part was outdated - it was written before feature about Buffer optimization.
Now the default value is one since we always need to have one buffer at least.
For example, if we take unet model with MatMul->Softmax->MatMul, there will be 3 Buffer with the same Reg Group -> `uniqie_buffer_reg_group_count = 1`

File: src/common/snippets/src/pass/mha_tokenization.cpp (hunk 2 above)
Comment (a-sidorova) [reply src/common/snippets/src/pass/mha_tokenization.cpp:271]: Because it's outdated code part. See: we check for `buffer_count < 2` however before we set `buffer_count = 2` - ??? - It seems like this lines are never executed
Moreover, as I wrote above, after MatMul0 we always have one Buffer -> so I set `uniqie_buffer_reg_group_count = 1` by default.
No another reason

File: src/common/snippets/src/pass/mha_tokenization.cpp (hunk 3 above)
Comment (a-sidorova) [reply src/common/snippets/src/pass/mha_tokenization.cpp:299]: Softmax also needs one Buffer but static Buffer in Softmax always reuse reg group of Buffer after MatMul0 -> no need to check static case here.

`MM0 -> Buffer [reg_group=0] -> .... Buffer in Softmax decomp [reg_group=0] ->....`

File: src/plugins/intel_cpu/src/transformations/transformation_pipeline.cpp (hunk 4 above)
Comment (a-sidorova) [reply src/plugins/intel_cpu/src/transformations/transformation_pipeline.cpp:870]: You're right!
But if dynamic Brgemm supports only fp32 in CPU, please move this condition on L878

File: src/tests/ov_helpers/ov_snippets_models/src/subgraph_mha.cpp [hunk 7]
```diff
@@ -269,58 +285,38 @@ std::shared_ptr<ov::Model> MHAMatMul0TransposeFunction::initReference() const {
...
+
+    const auto mulConst = ov::test::utils::make_constant(precisions[1], ov::Shape{1});
     const auto mul = std::make_shared<ov::op::v1::Multiply>(transpose1, mulConst);
-    const auto matMul0 = std::make_shared<ov::op::v0::MatMul>(transpose0, mul, transA, transB);
+    const auto matMul0 = std::make_shared<ov::op::v0::MatMul>(transpose0, mul);
     const auto add = std::make_shared<ov::op::v1::Add>(matMul0, addParam);
-    const auto reshape0 = std::make_shared<ov::opset1::Reshape>(add, reshape0Const, true);
-    const auto softMax = std::make_shared<ov::opset1::Softmax>(reshape0, 1);
-    const auto reshape1 = std::make_shared<ov::opset1::Reshape>(softMax, reshape1Const, true);
+    const auto softMax = std::make_shared<ov::opset1::Softmax>(add, rank - 1);
```
Comment (v-Golubev) [inline src/tests/ov_helpers/ov_snippets_models/src/subgraph_mha.cpp:313]: Now we always have softmax without reshapes in reference subgraph in spite of `with_reshape` value. Why? 
Was there previously a mistake in reference builder?

File: src/common/snippets/src/pass/mha_tokenization.cpp (hunk 5 above)
Comment (a-sidorova) [reply src/common/snippets/src/pass/mha_tokenization.cpp:313]: Because in general MatMul1 may have several consumers - however in this case `MatMul1` should the final node of the tokenized pattern.
Before this change we couldn't tokenize these patterns, although - why not?
```
    MatMul0
    Softmax
    MatMul1
    /    |     \
Op0  Op1  Op2
```

File: src/tests/ov_helpers/ov_snippets_models/src/subgraph_mha.cpp (hunk 7 above)
Comment (a-sidorova) [reply src/tests/ov_helpers/ov_snippets_models/src/subgraph_mha.cpp:313]: You're right!
Before there was mistake -> in CommonOptimization the pass ExtractReshapesAroundSoftmax removes these reshapes so reference subgraph shouldn't contain them,

File: src/common/snippets/src/pass/mha_tokenization.cpp (hunk 6 above)
Comment (a-sidorova) [reply src/common/snippets/src/pass/mha_tokenization.cpp:366]: If CI is red, I will fix in this PR. Otherwise I'd suggest to fix in the follow-up PR 😄

Comment (v-Golubev) [reply src/common/snippets/src/pass/mha_tokenization.cpp:366]: Of course)

File: src/tests/ov_helpers/ov_snippets_models/src/subgraph_mha.cpp (hunk 7 above)
Comment (v-Golubev) [reply src/tests/ov_helpers/ov_snippets_models/src/subgraph_mha.cpp:313]: Got it, thanks!

**Review by v-Golubev — APPROVED** [review]

**Comment by a-sidorova** [comment]
> * 123329

The results of functional validation has been attached to the ticket 123329

**Review by IvanNovoselov — APPROVED** [review]
